### PR TITLE
Add Graphql Query For The Current User

### PR DESCRIPTION
**Pull request by NickBolles**
My PR is a:
🚀 New feature

Main update on the:

Plugin - User-permissions

Add a GraphQL query for "User.me"

I need a little help on this. First off a question:
`How are the other user queries created?` (user, users etc)

Second, 
`How what's the correct resolver to include permissions for the query`
After reading the [GraphQL Guide](https://strapi.io/documentation/guides/graphql.html#apply-permissions-on-a-query), I thought using `resolver: 'User.me'` would do it, but that doesn't hit the controller method.

### CODE REVIEW

**Comment by NickBolles** [comment]
Looks like I missed a part of the docs (or they were updated in the last day or two???)

I think this is working now.

**Comment by kamalbennani** [comment]
Here you go, you figured it out by yourself, LGTM :rocket:

**Comment by danielmahon** [comment]
@kamalbennani @NickBolles this looks great, but are we exposing too much info here? I don't think you should be able to query all other users in your role.

Maybe we should remove the `users` field in the `User.me` controller's response?
https://github.com/strapi/strapi/blob/9b1c2c2b0d4048ac51c846381e61c5bdd2a3a246/packages/strapi-plugin-users-permissions/controllers/User.js#L43

This query lets me see emails and other info about other users
```
query {
  me {
    _id
    email
    role {
      _id
      name
      users {
        _id
        email
        username
      }
    }
  }
}
```

**Comment by NickBolles** [comment]
Interesting... Can you actually query those? I'm confused why you would be able to with graphql, but not the REST API. 

Can we confirm that we can actually query it and it's an issue with the REST API too?

**Comment by danielmahon** [comment]
@NickBolles You can query that with GraphQL but AFAIK the REST API doesnt return the `users` and `permissions` relation. Unless there is an additional query param or something to enable that. For GraphQL the `UserPermissionsRole` is where those are being defined:

```graphql
type UsersPermissionsRole {
  _id: ID!
  name: String!
  description: String
  type: String
  permissions(sort: String, limit: Int, start: Int, where: JSON): [UsersPermissionsPermission]
  users(sort: String, limit: Int, start: Int, where: JSON): [UsersPermissionsUser]
}
```

Im not sure how this should be fixed as it seems like maybe a bigger systemic problem of being able to run resolvers of nested relations without needing to pass a permissions check since you already passed the root one... maybe this is only a problem with the `me` route?

Im still super new to Strapi but seems like there are 3 options:
1. `me()` should return its own type (not `User`)
2. `/users/me` route needs to be locked down
3. nested relation resolvers should also pass permissions checks... complicated?

**Comment by NickBolles** [comment]
@danielmahon nice analysis! 

So my next question is do the normal `user` and `users` queries have this issue too? Essentially the `me` query is just a user query for the current user. 

I think the difference is how the ctx user is set up, which is in this file: https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-users-permissions/config/policies/permissions.js

But I think both use the find one controller here: https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-users-permissions/controllers/User.js

So maybe we can remove the roles users there. 

Also, are you sure you aren't querying as admin? I'm not sure how permissions should work with nested stuff like this. Someone more familiar with the `user-permissions` plugin will need to chime in on that.

**Comment by danielmahon** [comment]
@NickBolles `user` and `users` will indeed return the nested `role` and `users` but they should (IMO). And they are disabled by default. `me` however, is enabled by default and doesn't require authentication.

As a "Authenticated" user I can access `me`, but not `user` or `users` unless I add those routes to the permissions. As an "Administrator" user I can access everything.

I think the nested relation resolvers are being merged here:
https://github.com/strapi/strapi/blob/9b1c2c2b0d4048ac51c846381e61c5bdd2a3a246/packages/strapi-plugin-graphql/services/Resolvers.js#L271-L438

We could do a permissions check on every relation but I think thats going to get crazy complicated because you can't currently define the nested permissions, my gut tells me its not feasible.

I think the simplest solution would be to create a new return type for `me` =>`UsersPermissionsMe` and define `me.role` as JSON or flatten it, so that GraphQL doesnt see that as a relation and allow you to query it. Im not sure what that would break or how much is dependant on the current `/users/me` route.

I suppose it comes down to when you query for `me()` you aren't technically querying for `user()` so you shouldn't be given an actual `User` return type...

**Comment by kamalbennani** [comment]
@danielmahon As you said, this is really a big problem related to how GraphQL is implemented in Strapi, it's not just related to User.me, because we can easily have another model that has some relation with the user model or another model that an "Authenticated" user has access to and where he can easily have access to "nested" Data out of his scope. I think it's so hard to use nested permissions, I know that we've introduced something called "graphql-depth-limit" which limits the complexity of the GraphQL query by setting a maximum depth, but it'll not help us in this case.

**Comment by NickBolles** [comment]
It seems like the issue we are talking about is precisely what [graphql directives](https://www.apollographql.com/docs/graphql-tools/schema-directives.html#Enforcing-access-permissions) are for. 

If we had a field level directive that was `hasUserPermissionsRole` we could extend the permissions model to contain fields that this role doesn't have access to, and the `hasUserPermissionsRole` directive could look for the current field in any of the fields on the permissions model. 

I think we would already have all of this information  on the graphql context. 

This might be a bigger project than this PR though. Should I go with the custom return type for `me()` for now?

**Comment by danielmahon** [comment]
@NickBolles I vote for the custom return type for now, since the problem is GraphQL specific, but it's probably worth exploring directives in another PR.

File: packages/strapi-plugin-users-permissions/config/schema.graphql [hunk 1]
```diff
@@ -1,9 +1,26 @@
...
-  },  
+  },
+  definition: `
+    type UsersPermissionsMe {
+      _id: ID!
+      username: String!
+      email: String!
+      confirmed: Boolean
+      blocked: Boolean
+      role: UsersPermissionsRole
```
Comment (kamalbennani) [inline packages/strapi-plugin-users-permissions/config/schema.graphql]: Don't forget to use the new RoleType here: `UsersPermissionsMeRole`

**Pull request merged**